### ISSUE TITLE
[artys7-50] embedded flash size check always fails

### ISSUE THREAD

**Issue by majbthrd** [lowRISC/opentitan#4638]
Summary: when targeting the artys7-50 target, ADDR_MASK_EFLASH in tl_main_pkg.sv is auto-generated such that "opentitan_earlgrey_flash_size_check.py" stops the build.

Steps are:

```
$ . /tools/Xilinx/Vivado/2020.2/settings64.sh
$ cd $REPO_TOP
$ ./meson_init.sh
$ ninja -C build-out sw/device/boot_rom/boot_rom_export_fpga_nexysvideo
$ fusesoc --cores-root . run --flag=fileset_top --target=synth lowrisc:systems:top_earlgrey_artys7-50
$ ./hw/top_earlgrey/util/opentitan_earlgrey_flash_size_reduce.py 
$ fusesoc --cores-root . run --flag=fileset_top --target=synth lowrisc:systems:top_earlgrey_artys7-50
```

On the first run of fusesoc, the user is admonished and told to run ./hw/top_earlgrey/util/opentitan_earlgrey_flash_size_reduce.py 

```
INFO: Running pre_build script opentitan_earlgrey_flash_size_check
INFO:edalize.edatool:Running pre_build script opentitan_earlgrey_flash_size_check
ERROR: It seems that the size of the embedded flash has not been adjusted for the targeted FPGA device. The design might not fit. 
Please run hw/top_earlgrey/util/opentitan_earlgrey_flash_size_reduce.py before running this fusesoc core.
ERROR: Failed to build lowrisc:systems:top_earlgrey_artys7-50:0.1 : pre_build script 'opentitan_earlgrey_flash_size_check' exited with error code 1
ERROR:fusesoc.main:Failed to build lowrisc:systems:top_earlgrey_artys7-50:0.1 : pre_build script 'opentitan_earlgrey_flash_size_check' exited with error code 1
```

However, on the second run of fusesoc, despite the changes, the exact same failure happens.

Looking at the "opentitan_earlgrey_flash_size_check.py" that causes the failure, it is looking for both the RegPagesPerBank parameter in flash_ctrl_reg_pkg.sv as well as ADDR_MASK_EFLASH in tl_main_pkg.sv.  It the second that appears to cause the failure.

Even if I manually modify ADDR_MASK_EFLASH in tl_main_pkg.sv, the scripting just re-generates said file prior to the execution of "opentitan_earlgrey_flash_size_check.py".

**Comment by tjaychen** [comment]
thanks for bringing this up, i'll try to have a look.

**Comment by vogelpi** [comment]
Hi @majbthrd ,
thanks for reporting the issue (and thanks for pinging me @tjaychen ). I will have a look into this.

It's worth to mention that even if this issue is solved, the bitstream for the Artys7-50 probably can't be generated anymore. The reason is that in the past couple of weeks we have added some more blocks to the design which have high resource utilization on FPGA. As a result, we currently can't implement `top_earlgrey` on the smaller FPGAs. I have been working on PR to address this issue (#4329 ). The idea is to allow supporting multiple top configurations in the future. But off course, we won't have the bandwidth to keep supporting several boards. At the moment the main target for this PR is side channel analysis for the AES accelerator on FPGA where we also have to use a smaller FPGA. For the design to fit we need to remove some bigger blocks like KMAC, OTBN etc. 

For the Artys7-50, you have to ask yourself what exactly you need and then adjust the top-level as I did in #4329.

**Comment by vogelpi** [comment]
Alright, the problem is that with #4534 we updated the flash parameters:
- `PagesPerBank` changed from 256 -> 128
- `WordsPerPage` changed from 128 -> 256
This means the overall size stays the same.

But since the reduce script only changes `PagesPerBank` to 32, the flash size is now reduced only by a factor of 4 instead of 8. That's why the check for `ADDR_MASK_EFLASH` in `tl_main_pkg.sv` fails. It's now `0x1ffff` instead of `0xffff`.

I will create a PR to fix this. But the problem with FPGA size remains and I think this is also a separate issue.

**Comment by majbthrd** [comment]
Thanks, @vogelpi, the presence of the build files for the artys7-50 target, absence of any existing open/closed issues where others reported the build problem, and the provision for patching the flash size all made me mistakenly think that it much be viable.  It was only much later that I started to suspect that the present design was much too large for the artys7-50 after I saw the scale of FPGA utilization with the nexysvideo target.

I was intrigued from the onset of encountering this effort at the possibility of leveraging the OpenTitan scripting to target other (smaller) IBEX-based designs.  The learning curve is a bit steep, but your multiple top configuration PR gives some important clues on how that might be approached.

Thanks again.
